pkg/cmd/adm/restart.go

```diff
@@ -3,155 +3,198 @@ package adm
-	// get the deployment
-	deployment := &appsv1.Deployment{}
-	if err := cl.Get(context.TODO(), namespacedName, deployment); err != nil {
-		return 0, err
+func restartNonOlmDeployments(deployment appsv1.Deployment, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) error {
+
+	o := kubectlrollout.NewRolloutRestartOptions(ioStreams)
+
+	if err := o.Complete(f, nil, []string{"deployment"}); err != nil {
+		panic(err)
 	}
-	// keep original number of replicas so we can bring it back
-	originalReplicas := *deployment.Spec.Replicas
-	zero := int32(0)
-	deployment.Spec.Replicas = &zero
 
-	// update the deployment so it scales to zero
-	return originalReplicas, cl.Update(context.TODO(), deployment)
+	o.Resources = []string{"deployment/" + deployment.Name}
+
+	if err := o.Validate(); err != nil {
+		panic(err)
+	}
+	fmt.Printf("Running the rollout restart command for non-olm deployment %v", deployment)
+	return o.RunRestart()
 }
 
-func scaleBack(term ioutils.Terminal, cl runtimeclient.Client, namespacedName types.NamespacedName, originalReplicas int32) error {
-	return wait.Poll(500*time.Millisecond, 10*time.Second, func() (done bool, err error) {
-		term.Println("")
-		term.Printlnf("Trying to scale the deployment back to '%d'", originalReplicas)
-		// get the updated
-		deployment := &appsv1.Deployment{}
-		if err := cl.Get(context.TODO(), namespacedName, deployment); err != nil {
-			return false, err
-		}
-		// check if the replicas number wasn't already reset by a controller
-		if *deployment.Spec.Replicas == originalReplicas {
-			return true, nil
-		}
-		// set the original
-		deployment.Spec.Replicas = &originalReplicas
-		// and update to scale back
-		if err := cl.Update(context.TODO(), deployment); err != nil {
-			term.Printlnf("error updating Deployment '%s': %s. Will retry again...", namespacedName.Name, err.Error())
-			return false, nil
-		}
-		return true, nil
-	})
+func checkRolloutStatus(f cmdutil.Factory, ioStreams genericclioptions.IOStreams, labelSelector string) error {
+	cmd := kubectlrollout.NewRolloutStatusOptions(ioStreams)
+
+	if err := cmd.Complete(f, []string{"deployment"}); err != nil {
+		panic(err)
+	}
+	cmd.LabelSelector = labelSelector
+	if err := cmd.Validate(); err != nil {
+		panic(err)
+	}
+	fmt.Printf("Running the Rollout status to check the status of the deployment")
+	return cmd.Run()
+}
+
+func getExistingDeployments(cl runtimeclient.Client, ns string) (*appsv1.DeploymentList, *appsv1.DeploymentList, error) {
+
+	olmDeployments := &appsv1.DeploymentList{}
+	if err := cl.List(context.TODO(), olmDeployments,
+		runtimeclient.InNamespace(ns),
+		runtimeclient.MatchingLabels{"kubesaw-control-plane": "kubesaw-controller-manager"}); err != nil {
+		return nil, nil, err
+	}
+
+	nonOlmDeployments := &appsv1.DeploymentList{}
+	if err := cl.List(context.TODO(), nonOlmDeployments,
+		runtimeclient.InNamespace(ns),
+		runtimeclient.MatchingLabels{"provider": "codeready-toolchain"}); err != nil {
```

Comment:
also we could execute the listing only once , and get all the deployments in the namespace. Then filter them based on one of the labels only. 
Something like:

<details>

```
func getExistingDeployments(cl runtimeclient.Client, ns string) ([]appsv1.Deployment, []appsv1.Deployment, error) {
	var olmDeployments []appsv1.Deployment
	var nonOlmDeployments []appsv1.Deployment
	allDeployments := &appsv1.DeploymentList{}

	if err := cl.List(context.TODO(), allDeployments,
		runtimeclient.InNamespace(ns)); err != nil {
		return nil, nil, err
	}

	for _, deploy := range allDeployments.Items {
		if deploy.Labels != nil  {
			if _, found := deploy.Labels["kubesaw-control-plane"]; found {
				olmDeployments = append(olmDeployments, deploy)	
			}
		} else {
			nonOlmDeployments = append(nonOlmDeployments, deploy)
		}
	}
	
	return olmDeployments, nonOlmDeployments, nil
}
```

</details>

This way you limit the number of api calls and also rely on a single label. But feel free to leave it as is if you like it more.

Reply:
we do want to avoid a scenario where the non-olm deployments which are unrelated to kubesaw should not be  restarted. if a specific label is not provided this may heppen and hence i gave two label so that we are specific and limited too our deployments

Reply:
got it, I'm fine with performing 2 API calls and having 2 labels. thanks for explaining that.